crates/shared/src/price_estimation/competition.rs

```diff
@@ -771,71 +876,142 @@ mod tests {
+            OrderKind::Buy,
+            vec![
                 // User effectively pays `100_000` `sell_token`.
-                estimate(96_000, 1_000),
-                // User effectively pays `100_001` `sell_token`.
-                estimate(92_001, 2_000),
-            ),
-        ];
+                price(96_000, 1_000),
+                // User effectively pays `100_002` `sell_token`.
+                price(92_002, 2_000),
+                // User effectively pays `99_000` `sell_token` but the estimate
+                // gets discarded because it quotes 0 gas.
+                price(99_000, 0),
+            ],
+        )
+        .await;
+        assert_eq!(best, price(96_000, 1_000));
+    }
 
-        for (quote_kind, preferred_estimate, worse_estimate) in tests {
-            let priority: CompetitionEstimator<Arc<dyn PriceEstimating>> =
-                CompetitionEstimator::new(
-                    vec![vec![
-                        ("first".to_owned(), estimator(preferred_estimate)),
-                        ("second".to_owned(), estimator(worse_estimate)),
-                    ]],
-                    ranking.clone(),
-                );
-
-            let result = priority
-                .estimate(Arc::new(Query {
-                    kind: quote_kind,
-                    ..Default::default()
-                }))
-                .await;
-            assert_eq!(result.unwrap(), preferred_estimate);
-        }
+    /// If all estimators returned an error we return the one with the highest
+    /// priority.
+    #[tokio::test]
+    async fn returns_highest_priority_error() {
+        // Returns errors with highest priority.
+        let best = best_response(
+            PriceRanking::MaxOutAmount,
+            OrderKind::Sell,
+            vec![
+                error(PriceEstimationError::RateLimited),
+                error(PriceEstimationError::ProtocolInternal(anyhow::anyhow!("!"))),
+            ],
+        )
+        .await;
+        assert_eq!(best, error(PriceEstimationError::RateLimited));
+    }
+
+    /// Any price estimate, no matter how bad, is preferred over an error.
+    #[tokio::test]
+    async fn prefer_estimate_over_error() {
+        let best = best_response(
+            PriceRanking::MaxOutAmount,
+            OrderKind::Sell,
+            vec![
+                price(1, 1_000_000),
+                error(PriceEstimationError::RateLimited),
+            ],
+        )
+        .await;
+        assert_eq!(best, price(1, 1_000_000));
+    }
+
+    /// If all estimators returned an error we return the one with the highest
+    /// priority.
+    #[tokio::test]
+    async fn returns_highest_native_price() {
+        // Returns errors with highest priority.
+        let best = best_native_response(
+            PriceRanking::MaxOutAmount,
+            vec![native_price(1.), native_price(2.)],
+        )
+        .await;
+        assert_eq!(best, native_price(1.));
+    }
```

Comment:
🤔 

Is the assertion wrong here? How is `native_price(1.)` a higher native price than native_price(2.)?

Reply:
Yeah, this is indeed wrong. I got confused because native price estimates are `BUY` estimates where a lower `out_amount` is better. But we are denominating that number in ETH which makes bigger numbers better again.
A sanity check can be found [here](https://github.com/cowprotocol/services/pull/2195/files#diff-bf5c6f66e24401168ed4c2c36453fad49c5ee0f7ff88e6c845490bdeea5605bcL299-R297) where I replace:
`a < b` with `a.cmp(&b)` which is equivalent to `b >= a` what we had in the native case so I should also replace that with `a.cmp(&b)` as well.
Fixed and updated the test.